### PR TITLE
Use `br` instead of a conditional when switching on a constant boolean

### PR DESCRIPTION
**Pull request by clubby789**
r? @ghost

### CODE REVIEW

**Comment by clubby789** [comment]
Lets try applying this to all discriminants, not only bools
@bors try @rust-timer queue

**Comment by clubby789** [comment]
r? compiler

**Comment by fmease** [comment]
r? compiler

**Comment by michaelwoerister** [comment]
Looks promising but I don't know enough about this to gauge if this has any subtle, undesirable side effects.

r? compiler

File: compiler/rustc_codegen_ssa/src/mir/block.rs [hunk 1]
```diff
@@ -319,6 +319,14 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         targets: &SwitchTargets,
     ) {
         let discr = self.codegen_operand(bx, discr);
+        let discr_value = discr.immediate();
+        // If our discriminant is a constant we can branch directly
+        if let Some(const_discr) = bx.const_to_opt_u128(discr_value, true) {
```
Comment (compiler-errors) [inline compiler/rustc_codegen_ssa/src/mir/block.rs:324]: Why is `sign_ext` true?

Comment (clubby789) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:324]: I wasn't quite sure what the right option there was. Should it be `false` or depend on the type of the discr?

Comment (compiler-errors) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:324]: Yeah, I think it depends on the type of discr? I really don't know though -- I recommend looking at how these switch targets are evaluated.

Comment (clubby789) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:324]: I'll come back to it once perf has gone through for my two MIR-opt based approaches. They should allow us to waste less time emitting dead bbs to LLVM.

File: compiler/rustc_codegen_ssa/src/mir/block.rs [hunk 2]
```diff
@@ -319,6 +319,14 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         targets: &SwitchTargets,
     ) {
         let discr = self.codegen_operand(bx, discr);
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:321]: target_for_value is already being called below--the problem here is that it takes a u128, and consts are stored in the compiler as scalars in the MIR or machine-specific values here.

[the remark this answers is not in the document]
Comment (clubby789) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:321]: Would checking if it's a constant then `Const::try_to_bits` be sufficient?

**Comment by the8472** [comment]
Perfect timing, In #120682 I have some additional code paths behind `if CONST { impl_a() } else { impl_b() }` and it's blowing up some compile benchmarks. If this can fix that I'll have to do less IR-shaving.

File: compiler/rustc_codegen_ssa/src/mir/block.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (clubby789) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:321]: Looking into it, it looks like at this point we've already codegenned the `i1 true`, as `if true` creates a local which has `true` assigned to it (which wouldn't be safe to remove without checking for other uses of the local).
`codegen_operand` here then doesn't actually talk to LLVM, we go down to `maybe_codegen_consume_direct` which loads the `i1 true` reference directly from `self.locals`

**Comment by clubby789** [comment]
Reverted to the original version so I'm removing the regression label

**Comment by the8472** [comment]
To check my understanding for the purpose of optimizing library code:

Given

```rust
fn foo<T: SomeTrait, const C: bool>(x: T)
{
  if C {
    <T as SomeTrait>::a(x);
  } else {
    <T as SomeTrait>::b(x);
  }
}
```

Which work on the non-taken branch _and its call-tree_ would be avoided? Emitting llvmir, calculating generic `const {}` blocks, some other stuff?

**Comment by the8472** [comment]
Ah, that's less than I was looking for. The CopyProp one also wouldn't help in cases where the const only becomes known at monomorphization time, right?

File: compiler/rustc_codegen_ssa/src/mir/block.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:321]: > which loads the `i1 true` reference directly from `self.locals`

While that is true for `bool`, other types will need to generate an llvm value and load from it.

**Comment by saethlin** [comment]
r? saethlin

I'll review this by the end of the week.

**Comment by saethlin** [comment]
As opposed to all the other kinds of tweaks I want to make around lowering `SwitchInt(const _)` which may have knock-on effects such as not reporting post-mono errors, I'm pretty confident this change is perf-only.

@bors r+

**Comment by clubby789** [comment]
Looks like `sign_ext` indeed needs to be `false` - discriminants are stored sign-extended to their type size (https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/util/struct.Discr.html#structfield.val). Updated and added a new codegen test but this should probably be re-checked to make sure I'm right on this
@rustbot ready

**Comment by saethlin** [comment]
o7
@bors r+

**Pull request merged**